k8srm-prototype/pkg/api/claim_types.go

```diff
@@ -126,42 +74,95 @@ type DeviceClaim struct {
 // DeviceClaimSpec details the requirements that devices chosen
 // to satisfy this claim must meet.
 type DeviceClaimSpec struct {
-	// DeviceClass is the name of the DeviceClass containing the basic information
-	// about the device being requested.
+	// MatchAttributes allows specifying a constraint that will apply
+	// across all of the claims. For example, if you specified "numa", then
+	// this overall claim could only be successfully fulfilled if all of
+	// the included claims could be fulfilled by pools with the same "numa"
+	// attribute value. If we simply set matchAttributes in each claim
+	// separately, then they could be consistent within claims, but
+	// inconsistent across claims. Therefore, we need this additional
+	// constraint.
+	//
+	// +optional
+	MatchAttributes []string `json:"matchAttributes,omitempty"`
+
+	// Claims contains the actual claim details, arranged into groups
+	// containing claims which must all be satsified, or for which only
+	// one needs to be satisfied.
+	//
 	// +required
-	DeviceClass string `json:"deviceClass"`
+	Claims []DeviceClaimInstance `json:"claims,omitempty"`
+}
 
-	// Driver will limit the scope of devices considered to only those
-	// published by the specified driver. If the DeviceClass specifies a
-	// Driver, this should be left empty. If it is not, then it MUST match
-	// the Driver in the DeviceClass.
+// DeviceClaimInstance captures a claim which must be satisfied,
+// or a group for which one must be sastisfied.
+type DeviceClaimInstance struct {
+	// At least one of AllOf and OneOf must be populated.
+
+	// If fields of DeviceClaimDetail are populated, OneOf should
+	// be empty.
+	DeviceClaimDetail `json:",inline"`
```

Comment:
Should be a pointer - or we should make this an `allOf` array or we should give it a name like "claim".

Reply:
It's a little weird to embed a pointer and then use it as a OneOf.  I'll have to discuss with API reviewers.  Worst case, it becomes a named field.

Reply:
what is the difference of having a named field vs an array with just one element?

Reply:
I'm just trying to have one less nested stanza in the YAML...